Use 10 degree probing for BLTouch and check for errors

fredizzimo opened the pull request:
This also enables the BLTouch built in error checking. If there's a
fatal error the printer will shutdown. This is to prevent moving around
with the BLTouch probe in the down state.

Signed-off-by: Fred Sundvik <fsundvik@gmail.com>

**NOTE** I'm sending a pull request now, so that it can be tested. But I would like to add some further improvements. For example doing touch mode testing, if it's enabled by the configuration. The `pin_move_time` setting also needs to be much higher than before. I had unreliable error reporting with a move time of 500ms. Now I'm running with 1.5 seconds, but it could probably be smaller than that. I'm not sure how to handle a configuration change like that, should it be a warning, or a configuration error?

fredizzimo commented:
Actually now when I think about this, I don't think I can integrate reliable touch_mode tests into the actual probing. It would fail when you are too close to the bed.

So instead I could add a few commands to be used as config checks.
1. The first check is very basic, after the printer is turned on, the z-endstop should not be triggered. Even if the initial self-test fails, the failure should have been clear and the probe retracted.
2. The second check could be the touch_mode test (which might not work on all probes). That one is quite simple too. Enable touch_mode, move the probe up, and check that the endstop is triggered, then move the probe down, and verify that it isn't. After the test, reset the probe to normal mode
3. The third test could be a manual probe activation. Perform the same sequence as normal homing, but don't move the z-axis, instead let the user touch the probe
4. The fourth test should be a working homing procedure
5. The fifth test, tests the error handling. Homes the printer, then puts the pin down, and verifies that the error is triggered. This can be used to tune the `pin_move_time`

I think these tests should be implemented as a separate pull request, so in that sense this pull request is no longer WIP.

However, something still needs to be done about the new requirement for `pin_move_time` before it can be merged.

KevinOConnor commented:
Hi,

Let me start by saying that I do not own a BLTouch, and I'm happy to turn this over to someone who'd like to extend the BLTouch support!  However, we are about to make the next release of Klipper (the v0.7.0 release) and my priority right now is finding the source of the error reported in #955.  Any functionality improvements or additional error checks will have to wait until after the release.

Yes - catch errors in the g-code handler code and raise them as `raise gcode.error("my description")`.  The code will know it is invoked via g-code because it directly registers the command handler with gcode.register_command().

The only viable way to diagnose an mcu failure is with the Klipper log file.  The "rescheduled timer in the past" message is generally indicative of sending commands to the micro-controller with incorrect timing.  The original code did schedule commands with precise timing to ensure correct operation, but it is hard to say if it was the code changes that caused the error.

Cheers,
-Kevin

fredizzimo commented:
Hi @KevinOConnor 

I have checked the log that @J2xWo posted here https://github.com/KevinOConnor/klipper/issues/955#issuecomment-445571673. However I can't see any indication that events are sent in the wrong order yet there's the message "MCU 'mcu' shutdown: Rescheduled timer in the past". But then again, I might not be interpreting the logs the correct way.

As I see it, it went through the "connect" procedure just fine, issued the `reset` command followed by `pin_up`, then it went to prepare, which can be seen as log info messages, in addition to the send command dump. The prepare did a `pin_down`, which also successfully finished. Now the code moved on to the homing procedure, which I haven't modified at all.

It seems like it's starting the homing just fine, but somewhere in the middle there's a problem. The pattern changes at sent message 97, where it's no longer queuing steps with an interval of 250, it's also queuing 9 steps instead of the normal 7 ones. Another difference is that count stops being 65535. At the last step the intervals goes really high.

I don't see how my code could cause something like that, but it would be good to fix that so that @J2xWo can test the actual probing code. I can see that similar problems have been reported many times, and the usual conclusion has been that the speed is too fast for the MCU to handle.

I don't really buy that idea, at least not for the z-axis, which is configured to be just 10mm/s here. But in any case maybe @J2xWo could try lower it to maybe 5 and see if that helps?

And finally a question to @KevinOConnor, is it enough to call `printer.invoke_shutdown(msg)` before throwing the gcode exception to properly make it halt?

fredizzimo commented:
I will try your branch later today when I'm back home. I'll try both your version of the script, and a version where the `touch_mode` commands are disabled, to see if there are any differences.

BTW, I'm still struggling a bit to see how to easily add the gcode exception instead of the `mcu` one. I can't simply replace the current raise, because that would throw a gcode exception during the connect, which would be reported as an "Unhandled exception", with no further information. Additionally I don't think we are meant to raise gcode exceptions outside handling gcode.

I could catch the mcu exception in all relevant commands for homing and probing, but that means that I have to modify the code in many different places. And if more probing related commands are added in the future, it will very likely be forgotten.

I could catch the mcu exception somewhere in the `homing.py` file, and raise a gcode exception. Personally I don't like that either, since ´homing.py` does not know anything about gcode, and could just as well be called from somewhere else.

KevinOConnor commented:
Not sure I understand.  If the bltouch code detects a problem during its printer_state() method then catch it and throw a printer.config() error or pins.error().  If it detects a problem during a gcode handler then catch it and throw a gcode.error().  If it detects a problem during homing functions then catch it and throw a homing.EndStopError().

-Kevin

fredizzimo commented:
I could do that, but inside `bltouch.py` or `homing.py` how do I know that I'm inside a gcode handler?

fredizzimo commented:
@Elenedeath, since we are not allowed to discuss this in the other pull request, I'm asking you here.

Could you try to test my pull request again with just `G28`. From your latest test it seems like `touch_mode`, permanently puts the probe into a bad state, so that's something to avoid.

Elenedeath commented:
this is the result of a G28
[klippy.log](https://github.com/KevinOConnor/klipper/files/2750668/klippy.log)

fredizzimo commented:
Hm. It looked like it started to home properly, but never triggered.

How did it look like, did it start to move down? And where did it stop? Did it stop above the printbed, or did it actually hit it? Did it move far enough down, to make the pin jump up?

Elenedeath commented:
he move to endstop x and endstop Y and after this go to X150 and Y 150
bltouch pin down and stuck here

fredizzimo commented:
Ok, so the z-axis did not start to move? It really should have at that point. Also did you run the emergency stop command, or did the printer report the error itself?

I'm not really seeing any cause for the lack of movement, but it could be that you too have some incompatibility problems with the firmware on the controller board being newer than my pull request.

Elenedeath commented:
yeah z not moving and i stop with m112 after a 30 second wait.

fredizzimo commented:
I see that there was over 10 seconds between when the z-homing was supposed to start,and when the emergency stop was sent, so you probably did it automatically.

Since you seem to have two z-axis stepper motors, I find it even more strange that none of them moved. Yet your x and y axis moved properly.

There's one more thing you could try, and that is to manually trigger the bltouch when it's supposed to move down, by touching it (but be careful to not put finger grease on it, or to bend the pin when doing so, lightly moving your finger nail up should be fine).

After that the printer should "home". So now you could test if the standard `G1` commands works. If the z-axis still doesn't work, then there's definitely something strange with that.

Elenedeath commented:
i already try to push it when trying to home z and try to move z after and no move works.

yes i have a dual z stepper

fredizzimo commented:
Could you provide a log file, which shows me that? Also, after you manually touch the pin, does it stay up and turn red, or does it return back down?

Elenedeath commented:
[klippy.log](https://github.com/KevinOConnor/klipper/files/2750760/klippy.log)
this is a G28 with a G1 Z-10 and G1 Z50 but not on your branch

fredizzimo commented:
Thanks, but unfortunately that test is not valid, since his branch internally enables `touch_mode`, unless you have `triggered_on_pin_up` set to false. 

I also get slightly better logging from my own branch.

Elenedeath commented:
maybe i found the problem is about my z step distance with the help of kevin on irc

fredizzimo commented:
I have now rebased this on top of the current master and made some improvements and I no longer consider this as a work in progress. I think it's stable enough for inclusion

* The wire test is added back, and enabled by default, since it's useful in some situations
* The default pin_move time is set to 1.0 seconds
* Split the mcu error formatting to it's own commit

To test this pull request, run this:

```
sudo service klipper stop
git fetch origin pull/982/head:fredizzimo_bltouch_improvements
git checkout fredizzimo_bltouch_improvements
sudo service klipper start
```

And to return back to the master branch for example:

```
sudo service klipper stop
git checkout master
sudo service klipper start
```

@bshowket, I think this should work on your machine now

bshowket commented:
@fredizzimo thanks, I will move back to try.

bshowket commented:
@fredizzimo working your branch as expected, except i had to add test_sensor_pin: False in config. Thanks.

fredizzimo commented:
@markmosleh,

Could you please move your message over here, so that things don’t get confusing.


But I don’t think I can help you, unless you have a log file.

KevinOConnor commented:
Any further update on this pull request?

-Kevin

fredizzimo commented:
What exactly do I need to change? These are the things I'm aware of

1. Rebase on top of the latest
2. Since you now have fixed the endstop timing issues, I can change the endstop code reading code to use homing moves, like you do.
3. Always use the EndstopError exception instead of relying on `mcu.error` 
4. Move the changes to `gcode.py` to a new pull request, since it wouldn't have anything to do with this pull request anymore, without mcu.error exceptions. I think it's still nice to see the error messages in the console, rather than to have to dig through the logs. It doesn't affect any behaviour, since previously those have just flown through to the internal error case, and done the exactly same thing, including shutting the printer down. 

I would like to keep the shutdown of the printer, at least until we have proofs that it shuts down needlessly, even if there are no hardware errors. And in that case, we could add an option to disable it. I could also add the option immediately.

To handle the shutdown I could do the follwoing. 

1. At the startup, convert the exception to a config_error, like you do. Personally I would like to use `mcu.error`, because in the majority of the cases, we are talking about a hardware error, rather than a configuration error. But it's your choice.
2. Otherwise, create a new derivied exception `FatalEndstopError`. Handle this in `homing.py` and pass it further. The handling would do the following, if the axis has a `homing_retract_dist` configured, retract this first, then shutdown the printer. The reason for retracting, is to prevent the nozzle from touching the bed for too long if there's a hardware failure.

Is there anything else that would need to be changed?

danielkucera commented:
:D
